contracts/IdentityRegistryBridger.sol

```diff
@@ -0,0 +1,328 @@
+        if (newOwner == address(0)) {
+            revert ZeroAddress();
+        }
+
+        owner = newOwner;
+        emit OwnerUpdated(newOwner);
+    }
+
+    /// @dev Changes implementation contract address.
+    /// @notice Make sure implementation contract has function to change its implementation.
+    /// @param implementation Implementation contract address.
+    function changeImplementation(address implementation) external {
+        // Check for contract ownership
+        if (msg.sender != owner) {
+            revert OwnerOnly(msg.sender, owner);
+        }
+
+        // Check for zero address
+        if (implementation == address(0)) {
+            revert ZeroAddress();
+        }
+
+        // Store implementation address under designated storage slot
+        assembly {
+            sstore(PROXY_IDENTITY_REGISTRY_BRIDGER, implementation)
+        }
+        emit ImplementationUpdated(implementation);
+    }
+
+    /// @dev Registers or updates 8004 agent Id corresponding to service Id.
+    /// @param serviceId Service Id.
+    /// @return agentId Corresponding 8004 agent Id.
+    /// @return registered True if registered, otherwise updated.
+    function register(uint256 serviceId) external returns (uint256 agentId, bool registered) {
+        // Reentrancy guard
+        if (_locked > 1) {
+            revert ReentrancyGuard();
+        }
+        _locked = 2;
+
+        // Check for access
+        if (msg.sender != serviceManager) {
+            revert ManagerOnly(msg.sender, serviceManager);
+        }
+
+        // Get corresponding 8004 agent Id
+        agentId = mapServiceIdAgentIds[serviceId];
+
+        // TODO: if agent is decoupled - create new (as of now), or revert, or return false?
+        // Check existing 8004 agent for not being decoupled
+        if (agentId > 0) {
+            // Get agent Id owner
+            address agentOwner = IERC721(identityRegistry).ownerOf(agentId);
+            // Check for agent Id ownership
+            if (agentOwner != address(this)) {
+                // TODO Have external ownable function to check and decouple agents?
+                // Decouple current agent Id
+                emit AgentDecoupled(serviceId, agentId);
+
+                // Zero the agent Id such that the new one is created
+                agentId = 0;
+            }
+        }
+
+        // Check for 8004 agent Id correspondence
+        if (agentId == 0) {
+            registered = true;
+
+            // Create new agent Id
+            agentId = _register(serviceId);
+        } else {
+            // Update agent data, if applicable
+            _update(serviceId, agentId);
+        }
+
+        _locked = 1;
+    }
+
+    /// @dev Links service Ids with created 8004 agent Ids.
+    /// @param serviceIds Set of service Ids.
+    /// @return agentIds Set of 8004 agent Ids.
+    function link(uint256[] memory serviceIds) external returns (uint256[] memory agentIds) {
```

Comment:
We need to discuss whether we need to check some access here

Reply:
How are we going to link existing OLAS services?